src/docker_config.rs

```diff
@@ -0,0 +1,183 @@
+use base64::{engine::general_purpose, Engine as _};
+use oci_distribution::secrets::RegistryAuth;
+use serde::Deserialize;
+use std::collections::HashMap;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Error deserializing JSON: {0}")]
+    Deserializing(#[from] serde_json::Error),
+
+    #[error("Error decoding base64 field inside docker config auth section: {0}")]
+    Base64Decode(#[from] base64::DecodeError),
+
+    #[error("Error decoding content of base64 field inside docker config auth section: {0}")]
+    FromUtf8Error(#[from] std::string::FromUtf8Error),
+
+    #[error("Missing colon in auth field")]
+    MissingColon,
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// A content of ~/.docker/config.json file which is also the same format
+/// of the contents of the kubernetes kubernetes.io/dockerconfigjson secret
+#[derive(Clone, Deserialize)]
+pub struct DockerConfig {
+    auths: HashMap<String, DockerCredentials>,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+#[serde(untagged)]
+pub enum DockerCredentials {
+    Split { username: String, password: String },
+    Composite { auth: String },
+}
+
+impl DockerConfig {
+    /// Parse a DockerConfig from a string.
+    #[allow(dead_code)]
+    pub fn from_str(str: &str) -> Result<Self> {
+        Self::from_slice(str.as_bytes())
+    }
+
+    /// Parse a DockerConfig from slice of bytes.
+    pub fn from_slice(data: &[u8]) -> Result<Self> {
+        Ok(serde_json::from_slice(data)?)
+    }
+
+    /// Returns a [`RegistryAuth`] for a given image registry.
+    /// If a registry is not mentioned in the auth section of the docker config file,
+    /// the authentication method will be "anonymous" (i.e. unauthenticated), which
+    /// is suitable for public images. This matches the normal behavior of the docker client.
+    pub fn get_auth(&self, registry: &str) -> Result<RegistryAuth> {
+        Ok(match self.auths.get(registry) {
+            None => RegistryAuth::Anonymous,
+            Some(credentials) => {
+                let (username, password) = credentials.unpack()?;
+                RegistryAuth::Basic(username, password)
+            }
+        })
+    }
+}
+
+impl DockerCredentials {
+    fn unpack(&self) -> Result<(String, String)> {
+        Ok(match self.clone() {
+            DockerCredentials::Split { username, password } => (username, password),
+
+            DockerCredentials::Composite { auth } => {
+                String::from_utf8(general_purpose::STANDARD_NO_PAD.decode(auth)?)?
+                    .split_once(':')
+                    .map(|(a, b)| (a.to_string(), b.to_string()))
+                    .ok_or(Error::MissingColon)?
+            }
+        })
+    }
+}
+
+impl std::fmt::Debug for DockerConfig {
```

Comment:
note: we're not auto-deriving the `Debug` trait for the DockerConfig so that we don't accidentally leak secrets in logs.

The reasons it's still a good idea to manually implement the Debug trait:

1. so that someone (like my future self) doesn't accidentally add a derive(Debug) to it later
2. because `.expect_err` (used in tests) require it